### ISSUE TITLE
code canonical link function for poisson

### ISSUE THREAD

**Issue by hugoguh** [pavanramkumar/pyglmnet#42]
because it's the canonical, and for comparison with the non-canonical and for comparison with R

**Comment by pavanramkumar** [comment]
I will modify the class to include an additional parameter called `link` which can be either `softplus` (the current one) or `exp` (canonical). I will take this up after issue #37 is closed.

**Comment by pavanramkumar** [comment]
I added a separate `distr` for the canonical Poisson targets called `poissonexp`. 

However, the gradients are unstable due to overflow of `numpy.exp()` and `fit()` returns `nan` for all coefficients.

I believe that the R and Matlab implementations of `glmnet` also face this issue for the `poisson` case and they probably use some modifications of `exp()` to prevent over- or under-flow.

I can look into this more carefully later, but for now, I will exclude it from the first release milestone.

**Comment by hugoguh** [comment]
Have you tried using an optimizer, let's see is an optimizer finds the solution when we give it the cost function.
Does it consistently fail? Even with low firing neurons and good initializations?

**Comment by pavanramkumar** [comment]
I have managed to get the canonical link `poissonexp` to work. I use a simple hard threshold to prevent overflow. However, the way I currently implement the nonlinearity `qu` is as follows:

```python
qu = np.exp([zi if zi < 5.0 else 5.0 for zi in z.ravel()])
```
This for loop seems to be very slow and makes `10000` x `100` problems extremely slow. Anyone know a faster way to do this?

**Comment by hugoguh** [comment]
How about making it a smooth transition between `np.exp` and `np.log(1+exp(x))`
having as relative weight a `logistic` that crosses zero at `5` or whatever threshold you find good

**Comment by hugoguh** [comment]
something like this

```python
thres = 5

plt.figure(figsize=(10,5))
x = np.linspace(-2,12,100)

plt.subplot(121)
weight = 1/(1+np.exp(thres-x))
plt.plot(x, weight)
plt.title('shifted logistic')

plt.subplot(122)
func = (1-weight)*np.exp(x)+(weight)*np.log(1+np.exp(x))
plt.plot(x, np.exp(x))
plt.plot(x, func)
plt.ylim([0,1000])
plt.legend(['np.exp','new'],loc=2)
plt.show()
```
![screen shot 2016-05-10 at 1 33 34 pm](https://cloud.githubusercontent.com/assets/3664942/15158068/d547fd10-16b3-11e6-8eb0-33f7e87e569f.png)

you can change the steepness of the logistic also to make a more sharp but still smooth transition

**Comment by hugoguh** [comment]
maybe a good time to try a couple of these optimizers?
http://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.optimize.minimize.html

To check if any of them manages to avoid this problem. And also to test the convergence of our gradient descent implementation where it seems to be working

**Comment by pavanramkumar** [comment]
ok, this is significantly faster and works well for `10000` x `100`:
```python
z[z > 5.0] = 5.0
qu = np.exp(z)
```

**Comment by pavanramkumar** [comment]
@hugoguh i will try the optimizers, but i need to hand-compute the hessians for each distribution first. i will do that. follow the updates in issue #39

**Comment by hugoguh** [comment]
It’s a bit ugly though no? Why 5? How will it affect the fits
we should test the `scipy.optimize` and see if any optimizer manages to avoid it and try to implement it

**Comment by hugoguh** [comment]
Ah ok, only saw your reply now.
I think some don’t need hessian, but with hessian you can try more, and also expect better performance.
Cool

**Comment by pavanramkumar** [comment]
this issue is now closed by commit 135f488c3b3f343bd4d06141ce41b51d7bd054cf

**Comment by jasmainak** [comment]
@pavanramkumar tests would be nice :)
